### ISSUE TITLE
 tmp-submul_1.s:110: Error: selected processor does not support ARM mode `umaal r6,r12,r4,r3'

### ISSUE THREAD

**Issue by 8ohmh** [crosstool-ng/crosstool-ng#857]
Hi,

got several times: (Debian latest,  crosstool-ng-1.23.0, chroot host)
```
[ALL  ]    tmp-addmul_1.s: Assembler messages:
[ALL  ]    tmp-addmul_1.s:102: Error: selected processor does not support ARM mode `umaal r6,r12,r4,r3'
[ALL  ]    tmp-addmul_1.s:106: Error: selected processor does not support ARM mode `umaal r7,r12,r5,r3'
[ALL  ]    tmp-addmul_1.s:110: Error: selected processor does not support ARM mode `umaal r6,r12,r4,r3'
[ALL  ]    tmp-addmul_1.s:114: Error: selected processor does not support ARM mode `umaal r7,r12,r5,r3'
[ALL  ]    tmp-addmul_1.s:120: Error: selected processor does not support ARM mode `umaal r6,r12,r4,r3'
[ALL  ]    Makefile:768: recipe for target 'addmul_1.lo' failed
[ERROR]    make[3]: *** [addmul_1.lo] Error 1
[ALL  ]     armv7l-build_unknown-linux-gnueabi-gcc -std=gnu99 -c -DHAVE_CONFIG_H -I. -I/home/user/crosstools/crosstool-ng-1.23.0/.build/src/gmp-6.1.2/mpn -I.. -D__GMP_WITHIN_GMP -I/home/user/crosstools/crosstool-ng-1.23.0/.build/src/gmp-6.1.2 -DOPERATION_submul_1 -O2 -g -pipe -I/home/user/crosstools/crosstool-ng-1.23.0/.build/arm-unknown-eabi/buildtools/include -fexceptions -Wa,--noexecstack tmp-submul_1.s -o submul_1.o
[ALL  ]    tmp-submul_1.s: Assembler messages:
[ALL  ]    tmp-submul_1.s:110: Error: selected processor does not support ARM mode `umaal r6,r12,r4,r3'
[ALL  ]    tmp-submul_1.s:116: Error: selected processor does not support ARM mode `umaal r7,r12,r5,r3'
[ALL  ]    tmp-submul_1.s:121: Error: selected processor does not support ARM mode `umaal r6,r12,r4,r3'[ALL  ]    tmp-addmul_1.s: Assembler messages:
[ALL  ]    tmp-addmul_1.s:102: Error: selected processor do[ALL  ]    tmp-addmul_1.s: Assembler messages:
[ALL  ]    tmp-addmul_1.s:102: Error: selected processor does not support ARM mode `umaal r6,r12,r4,r3'
[ALL  ]    tmp-addmul_1.s:106: Error: selected processor does not support ARM mode `umaal r7,r12,r5,r3'
[ALL  ]    tmp-addmul_1.s:110: Error: selected processor does not support ARM mode `umaal r6,r12,r4,r3'
[ALL  ]    tmp-addmul_1.s:114: Error: selected processor does not support ARM mode `umaal r7,r12,r5,r3'
[ALL  ]    tmp-addmul_1.s:120: Error: selected processor does not support ARM mode `umaal r6,r12,r4,r3'
[ALL  ]    Makefile:768: recipe for target 'addmul_1.lo' failed
[ERROR]    make[3]: *** [addmul_1.lo] Error 1
[ALL  ]     armv7l-build_unknown-linux-gnueabi-gcc -std=gnu99 -c -DHAVE_CONFIG_H -I. -I/home/user/crosstools/crosstool-ng-1.23.0/.build/src/gmp-6.1.2/mpn -I.. -D__GMP_WITHIN_GMP -I/home/user/crosstools/crosstool-ng-1.23.0/.build/src/gmp-67cefe06f1359da78f559e81c6c1f46994a928d64  umaal-error.tar.gz
.1.2 -DOPERATION_submul_1 -O2 -g -pipe -I/home/user/crosstools/crosstool-ng-1.23.0/.build/arm-unknown-eabi/buildtools/include -fexceptions -Wa,--noexecstack tmp-submul_1.s -o submul_1.o
[ALL  ]    tmp-submul_1.s: Assembler messages:
[ALL  ]    tmp-submul_1.s:110: Error: selected processor does not support ARM mode `umaal r6,r12,r4,r3'
[ALL  ]    tmp-submul_1.s:116: Error: selected processor does not support ARM mode `umaal r7,r12,r5,r3'
[ALL  ]    tmp-submul_1.s:121: Error: selected processor does not support ARM mode `umaal r6,r12,r4,r3'7cefe06f1359da78f559e81c6c1f46994a928d64  umaal-error.tar.gz

[ALL  ]    tmp-submul_1.s:126: Error: selected processor does not support ARM mode `umaal r7,r12,r5,r3'
[ALL  ]    tmp-submul_1.s:133: Error: selected processor does not support ARM mode `umaal r6,r12,r4,r3'
es not support ARM mode `umaal r6,r12,r4,r3'
[ALL  ]    tmp-addmul_1.s:106: Error: selected processor does not support ARM mode `umaal r7,r12,r5,r3'
[ALL  ]    tmp-addmul_1.s:110: Error: selected processor does not support ARM mode `umaal r6,r12,r4,r3'
[ALL  ]    tmp-addmul_1.s:114: Error: selected processor does not support ARM mode `umaal r7,r12,r5,r3'
[ALL  ]    tmp-addmul_1.s:120: Error: selected processor does not support ARM mode `umaal r6,r12,r4,r3'
[ALL  ]    Makefile:768: recipe for target 'addmul_1.lo' failed
[ERROR]    make[3]: *** [addmul_1.lo] Error 1
[ALL  ]     armv7l-build_unknown-linux-gnueabi-gcc -std=gnu99 -c -DHAVE_CONFIG_H -I. -I/home/user/crosstools/crosstool-ng-1.23.0/.build/src/gmp-6.1.2/mpn -I.. -D__GMP_WITHIN_GMP -I/home/user/crosstools/crosstool-ng-1.23.0/.build/src/gmp-6.1.2 -DOPERATION_submul_1 -O2 -g -pipe -I/home/user/crosstools/crosstool-ng-1.23.0/.build/arm-unknown-eabi/buildtools/include -fexceptions -Wa,--noexecstack tmp-submul_1.s -o submul_1.o
[ALL  ]    tmp-submul_1.s: Assembler messages:
[ALL  ]    tmp-submul_1.s:110: Error: selected processor does not support ARM mode `umaal r6,r12,r4,r3'
[ALL  ]    tmp-submul_1.s:116: Error: selected processor does not support ARM mode `umaal r7,r12,r5,r3'
[ALL  ]    tmp-submul_1.s:121: Error: selected processor does not support ARM mode `umaal r6,r12,r4,r3'
[ALL  ]    tmp-submul_1.s:126: Error: selected processor does not support ARM mode `umaal r7,r12,r5,r3'
[ALL  ]    tmp-submul_1.s:133: Error: selected processor does not support ARM mode `umaal r6,r12,r4,r3'

[ALL  ]    tmp-submul_1.s:126: Error: selected processor does not support ARM mode `umaal r7,r12,r5,r3'
[ALL  ]    tmp-submul_1.s:133: Error: selected processor does not support ARM mode `umaal r6,r12,r4,r3'
```
Attached config.log, build.log etc:
[umaal-error.tar.gz]
(https://github.com/crosstool-ng/crosstool-ng/files/1446125/umaal-error.tar.gz)
`SHA1 7cefe06f1359da78f559e81c6c1f46994a928d64  umaal-error.tar.gz`

[dmesg.tar.gz](https://github.com/crosstool-ng/crosstool-ng/files/1446138/dmesg.tar.gz)
9baa4f63d8cd39b013852fc842b2973c1ba773a3  ./dmesg.tar.gz

**Comment by 8ohmh** [comment]
Can't even compile x86_64. Tried even to pach .config CT_GMP=n etc. but doesn't work (MPFR is depending on): Host compiler says:
```
Using built-in specs.
COLLECT_GCC=gcc
COLLECT_LTO_WRAPPER=/usr/lib/gcc/arm-linux-gnueabi/4.9/lto-wrapper
Target: arm-linux-gnueabi
Configured with: ../src/configure -v --with-pkgversion='Debian 4.9.2-10' --with-bugurl=file:///usr/share/doc/gcc-4.9/README.Bugs --enable-languages=c,c++,java,go,d,fortran,objc,obj-c++ --prefix=/usr --program-suffix=-4.9 --enable-shared --enable-linker-build-id --libexecdir=/usr/lib --without-included-gettext --enable-threads=posix --with-gxx-include-dir=/usr/include/c++/4.9 --libdir=/usr/lib --enable-nls --with-sysroot=/ --enable-clocale=gnu --enable-libstdcxx-debug --enable-libstdcxx-time=yes --enable-gnu-unique-object --disable-libitm --disable-libquadmath --enable-plugin --with-system-zlib --disable-browser-plugin --enable-java-awt=gtk --enable-gtk-cairo --with-java-home=/usr/lib/jvm/java-1.5.0-gcj-4.9-armel/jre --enable-java-home --with-jvm-root-dir=/usr/lib/jvm/java-1.5.0-gcj-4.9-armel --with-jvm-jar-dir=/usr/lib/jvm-exports/java-1.5.0-gcj-4.9-armel --with-arch-directory=arm --with-ecj-jar=/usr/share/java/eclipse-ecj.jar --enable-objc-gc --enable-multiarch --disable-sjlj-exceptions --with-arch=armv4t --with-float=soft --enable-checking=release --build=arm-linux-gnueabi --host=arm-linux-gnueabi --target=arm-linux-gnueabi
Thread model: posix
gcc version 4.9.2 (Debian 4.9.2-10) 


```

**Comment by 8ohmh** [comment]
So far I found the source for the problem; The installed Debian compiler which is used for build doesn't support arm v7, cortex a9. 
1. Normally the compiler doesn't should care about that)
2. I am wondering why I get on my x86_64 debian an compiler with arm-v7 support but not on my arm machine

**Comment by stilor** [comment]
First, as our issue submission guidelines suggest, please run `ct-ng build.1`. It is hard to read a log of multiple parallel jobs.

Second, I think config.log from GMP's build directory (`.build/arm-unknown-eabi/build/build-gmp-host-armv7l-build_unknown-linux-gnueabi`) may have some clues, please attach it as well.

But ultimately, I think this is more of a question to GMP mailing lists - I may not be able to help you much.

**Comment by 8ohmh** [comment]
Thanks for the answer.

1. I've already dropped an issue to the mail of gmp project, but till now 
got no answer - seems that it is a bigger problem

2. Although thath both Debian versions are the same (on Board and on PC), they
got different multilib/march ??????

3. I tried all possible with-march, tune for cpu, so I would like to 
have a feedback in Crosstools-NG which (env, etc) vars are used short BEFORE 
GMP is called (a simple export command). How to do that (and where?) 
before I guessing around insode of Crosstools-NG

4. Maybe I'm doing a mistake:

How can I setup Crosstools-NG with my x86_64 Debian Linux to get a 
toolchain for armel V7, cortex-a9 - and/or ARMv6T2 / Cortex-a8. Beside 
of the original compiler I have also a board manufacturer provided gcc 
residing in /home/user/special_arm_compiler (the special_compiler-gcc 
-as, -ld etc. residing in that /home/user/special_arm_compiler/bin/ folder.
How to setup HOST/CANADIAN CROSS NATIVE?

I just want to compile a toolchain which is running on that arm board 
with debian linux. What host, build etc., arch, prefix do I have to 
use?

**Comment by stilor** [comment]
2. That's expected given that the list of multilibs/multiarch depends on the architecture. AFAIU you have an x86 PC and an ARM board, right?

3. In the build log, the full environment is dumped. But that prints the environment common to all steps; specific steps (such as GMP) may modify it further; you'll need to look into the respective build scripts (e.g. scripts/build/companion_libs/100-gmp.sh) for any env vars set on top of that.

4. ARM is notorious in that it does not allow one to set a specific list of multilibs to build, so for a single toolchain to compile for different targets, you'll likely need to modify GCC sources. So most likely you'll need two separate toochains for armv7 and armv6.

I suggest you start with one of the samples, e.g. `arm-cortexa9_neon-linux-gnueabihf`. Does it build for you? If it does, can you verify that it produces executables that can run on your board?

Once we're past that step, we can then try to use the result of that build to produce a canadian configuration to build a cross-native toolchain.

**Comment by 8ohmh** [comment]
To 2: I have an x86_64 DEBIAN PC and that ARM Board - on which beneath a simple ARM Linux another DEBIAN Linux chrooted runs (with the known problems). I tried to compile the toolchain on that (!!!) board under the chrooted Debian, but it is always falining - even compiling an toolchain arm->x86-i386  

To 3: I will try to compile it again on that board with the improvements you wrote me: 
a. single instance of compiling
b. putting all logs in to tar.gz file
c. I just was wondering that - although I have the same Debian versions running on both machines, the ARM board version has a compiler which is different in multilib/march and compiling arm (beside it should give both the same arm code). This is also one thing I'm wondering.

Already I got a cross version by compiling on the x86 machine and I got special arm gcc by manufacturer provided but the latter hangs in integrating into crosstools-ng

**Comment by 8ohmh** [comment]
BTW Many thanks for the answers.

**Comment by 8ohmh** [comment]
Here are (by Stilor demanded) logs:

[crosstools-ng_gmp_failing.tar.gz](https://github.com/crosstool-ng/crosstool-ng/files/1484914/crosstools-ng_gmp_failing.tar.gz)

9a5d4489fd80c8fc9fcb5b7107836948  ./crosstools-ng_gmp_failing.tar.gz

**Comment by 8ohmh** [comment]
Solved the problem: crosstools was using gcc-4.4 instead of installed gcc-6, so it could compile

**Comment by 8ohmh** [comment]
Sorry, could solve the problem - Partly my fault. I was expecting that Crosstools-NG doing a make clean before any new start and I dropped in CFlags extra an option "-mcpu=marvell-pj4". Now it compiles (after a ct-ng clean) till beyond GMP
